### PR TITLE
Enable the QUIC demos to take hostname/port from the command line

### PR DESCRIPTION
**Pull request by mattcaswell**
We amend the various QUIC demos from the guide to be able to accept hostname/port from the command line.

We also significantly extend the README to provide more information about how to run the demos, and how to get a server that they can connect to.

This address these project issues:

Fixes openssl/project#252
Fixes openssl/project#261

Note that openssl/project#252 says that ALPN should also be provided via the command line. However, while implementing this I realised that this makes little sense. The demos are fixed to use HTTP/1.0 and cannot support any other protocol - so it doesn't make much sense to be able to supply something else on the command line.

### CODE REVIEW

**Review by t8m — APPROVED** [review]

File: demos/guide/README.md [hunk 1]
```diff
@@ -5,11 +5,73 @@ The demos in this directory are the complete source code for the applications
...
+Running the TLS Demos
+---------------------
+
 To run the demos when linked with a shared library (default) ensure that
 libcrypto and libssl are on the library path. For example, assuming you have
 already built OpenSSL from this source and in the default location then to run
 the tls-client-block demo do this:
 
-LD_LIBRARY_PATH=../.. ./tls-client-block
+LD_LIBRARY_PATH=../.. ./tls-client-block hostname port
```
Comment (paulidale) [inline demos/guide/README.md:19]: Shouldn't we be recommending util/wrap.pl for this?

File: demos/guide/README.md [hunk 2]
```diff
@@ -5,11 +5,73 @@ The demos in this directory are the complete source code for the applications
...
+
+The above assumes that your default trusted certificate store containing trusted
+CA certificates has been properly setup and configured as described on the
+[TLS Introduction] page.
+
+You can run a test server to try out these demos using the "openssl s_server"
+command line utility and using the test server certificate and key provided in
+this directory. For example:
+
+openssl s_server -www -accept localhost:4443 -cert servercert.pem -key serverkey.pem
```
Comment (paulidale) [inline demos/guide/README.md:32]: Should this be wrapped?  `openssl` will pick up the system version if present and nothing if not.

File: doc/man7/ossl-guide-tls-client-non-block.pod [hunk 3]
```diff
@@ -252,13 +252,26 @@ retry. An optional mode does exist (B<SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER>)
 which will configure OpenSSL to allow the buffer being written to change from
 one retry to the next. However, in this case, you must still retry exactly the
 same data - even though the buffer that contains that data may change location.
-See L<SSL_CTX_set_mode(3)> for further details.
+See L<SSL_CTX_set_mode(3)> for further details. As before we write the request
```
Comment (paulidale) [inline doc/man7/ossl-guide-tls-client-non-block.pod:255]: There is no _before_ in this document.  _As in the TLS tutorials_ was the wording used elsewhere, although I don't really see the need to reference the TLS tutorials but if they are referenced, it should be a link IMO.

**Review by paulidale — CHANGES_REQUESTED** [review]

**Review by levitte — APPROVED** [review]

File: demos/guide/README.md (hunk 1 above)
Comment (mattcaswell) [reply demos/guide/README.md:19]: I'm not sure that we should. `util/wrap.p` sets up a whole load of other environment variables (e.g. `OPENSSL_CONF`, `OPENSSL_CONF_INCLUDE` etc) that are specifically intended for us during our own testing. I see `util/wrap.pl` as something mostly internal. I'm not sure we want to be encouraging end users to use it?

File: demos/guide/README.md (hunk 2 above)
Comment (mattcaswell) [reply demos/guide/README.md]: I added `LD_LIBRARY_PATH` etc to make this work

**Comment by mattcaswell** [comment]
Updated to address feedback.

@t8m, @levitte: please reconfirm

File: demos/guide/README.md [hunk 4]
```diff
@@ -29,7 +29,7 @@ You can run a test server to try out these demos using the "openssl s_server"
 command line utility and using the test server certificate and key provided in
 this directory. For example:
 
-openssl s_server -www -accept localhost:4443 -cert servercert.pem -key serverkey.pem
+LD_LIBRARY_PATH=../.. ../../apps/openssl s_server -www -accept localhost:4443 -cert servercert.pem -key serverkey.pem
```
Comment (t8m) [inline demos/guide/README.md:32]: I am unsure this is better, but anyway.

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by hlandau** [comment]
Merged to master and 3.2. Thank you.

**Pull request closed without merging**